File: DSA-3/Queues/lru-cache.cpp
/******

https://www.geeksforgeeks.org/lru-cache-implementation/

*****/
#include<bits/stdc++.h>
using namespace std;
class LRUCache
{
public:
    LRUCache(int );
 
    int get(int );
 
    void set(int , int );
};
int main()
{
    int t;
    cin>>t;
    while(t--)
    {
    int n;
    cin>>n;
    LRUCache *l  = new LRUCache(n);
    int q;
    cin>>q;
    for(int i=0;i<q;i++)
    {
        string a;
        cin>>a;
        if(a=="SET")
        {
            int aa,bb;
            cin>>aa>>bb;
            l->set(aa,bb);
        }else if(a=="GET")
        {
            int aa;
            cin>>aa;
            cout<<l->get(aa)<<" ";
        }
    }
    cout<<endl;
    }
}

/*This is a function problem.You only need to complete the function given below*/
/*The structure of the class is as follows 
class LRUCache
{
public:
    LRUCache(int );
    int get(int );
    void set(int , int );
};*/
/*You are required to complete below methods */
/*Inititalize an LRU cache with size N */
int maxSize;
list<pair<int,int>> dq;
unordered_map<int, list<pair<int,int>> :: iterator> ma;

LRUCache::LRUCache(int N){
     //Your code here
     maxSize = N;
     dq.clear();
     ma.clear();
     
}
/*Sets the key x with value y in the LRU cache */
void LRUCache::set(int x, int y) {
     //Your code here
     if(ma.find(x) == ma.end()){
         
         if(ma.size() == maxSize){
             
             int first = dq.back().first;
             int second = dq.back().second;
             dq.pop_back();
             dq.push_front({x,y});
             ma.erase(first);
             ma[x] = dq.begin();
        }else{
             dq.push_front({x,y});
             ma[x] = dq.begin();
         }
     }else{
        auto it = ma.find(x);
        dq.erase(it->second);
        dq.push_front({x,y});
   
        ma[x]=dq.begin();
     }
}
/*Returns the value of the key x if 
present else returns -1 */
int LRUCache::get(int x){
    //Your code here
    if(ma.find(x) == ma.end()){
        return -1;
    }
    
    auto it = ma.find(x);
    int d = (*(it->second)).second;
    dq.erase(it->second);
    dq.push_front({x,d});
    ma[x] = dq.begin();
    return d;
}

File: Company Wise Practice/Goldman Sachs/lru-cache.java

import java.util.Scanner;
import java.util.*;
class LRU_Cache
{
	public static void main(String args[])
	{
		Scanner sc = new Scanner(System.in);
		int t = sc.nextInt();
		while(t>0)
		{
			Set<Integer> s = new HashSet<Integer>() ;
				int n = sc.nextInt();
				LRUCache g = new LRUCache(n);
			int q = sc.nextInt();
			
			while(q>0)
			{
			
				String c = sc.next();
			//	System.out.println(c);
				if(c.equals("GET"))
				{
					int x = sc.nextInt();
					System.out.print(g.get(x)+" ");
				}
				if(c.equals("SET"))
				{
					int x = sc.nextInt();
					int y  = sc.nextInt();
					g.set(x,y);
				}
			
			q--;
			//System.out.println();
			}
		t--;
		System.out.println();
		}
	}
}


/*This is a function problem.You only need to complete the function given below*/
/*You are required to complete below class */
class LRUCache {
 
   Map<Integer,Integer> map;
    int capacity;
    Queue<Integer> q;
    
    /*Inititalize an LRU cache with size N */
    public LRUCache(int N) {
       //Your code here
       capacity = N;
       map = new HashMap<>(N);
       q = new LinkedList<>();
       
    }
    
    /*Returns the value of the key x if 
     present else returns -1 */
    public int get(int x) {
       //Your code here
       if(map.containsKey(x)){
           q.remove(new Integer(x));
           q.add(x);
           return map.get(x);
       }
        
        return -1;
       
    }
    
    /*Sets the key x with value y in the LRU cache */
    public void set(int x, int y) {
       //Your code here
       if(!q.contains(x)){
           
           if(q.size() == capacity){
               int oldest = q.poll();
               map.remove(oldest);
           }
       }else{
           q.remove(new Integer(x));
       }
       
       q.add(x);
       map.put(x,y);
    }
}
